### PR TITLE
integration (functional) tests for web API project

### PR DESCRIPTION
**Pull request by SergKanz**

### CODE REVIEW

File: test/WebApplication1/Controllers/ValuesController.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
+﻿using System;
```
Comment (olegsych) [inline test/WebApplication1/Controllers/ValuesController.cs]: Is this and other WebApplication1 files duplicates of SampleWebAPIIntegration?

File: test/SampleWebAPIIntegration/Project_Readme.html [hunk 2]
```diff
@@ -0,0 +1,203 @@
+﻿<!DOCTYPE html>
```
Comment (olegsych) [inline test/SampleWebAPIIntegration/Project_Readme.html]: Do we need this README file?

File: test/SampleWebAPIIntegration/Controllers/ValuesController.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        // GET: api/values
+        [HttpGet]
+        public IEnumerable<string> Get()
+        {
+            return new string[] { "value1", "value2" };
+        }
+
+        // GET api/values/5
+        [HttpGet("{id}")]
+        public string Get(int id)
```
Comment (olegsych) [inline test/SampleWebAPIIntegration/Controllers/ValuesController.cs]: Do we need any methods except Get?

File: test/SampleWebAPIIntegration/SampleWebAPIIntegration.xproj [hunk 4]
```diff
@@ -0,0 +1,22 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (olegsych) [inline test/SampleWebAPIIntegration/SampleWebAPIIntegration.xproj]: This project contains what we call functional tests. Do we want the name of the project to reflect that? WebAPI.FunctionalTests perhaps?

File: test/SampleWebAPIIntegration/FunctionalTest/RequestTelemetryTests.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
+﻿namespace SampleWebAppIntegration.FunctionalTest
+{
+    using FunctionalTestUtils.Tests;
+    using Microsoft.ApplicationInsights.DataContracts;
+    using System.Linq;
+    using Xunit;
+
+    public class RequestTelemetryTests : RequestTelemetryTestsBase
```
Comment (olegsych) [inline test/SampleWebAPIIntegration/FunctionalTest/RequestTelemetryTests.cs]: Can we talk about value of these tests now ? :smiley:

File: test/SampleWebAPIIntegration/Startup.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        // Use this method to add services to the container
+        public void ConfigureServices(IServiceCollection services)
+        {
+            // Add Application Insights services to the services container.
+            services.AddApplicationInsightsTelemetry(Configuration);
+
+            services.AddMvc();
+            // Uncomment the following line to add Web API services which makes it easier to port Web API 2 controllers.
+            // You will also need to add the Microsoft.AspNet.Mvc.WebApiCompatShim package to the 'dependencies' section of project.json.
+            // services.AddWebApiConventions();
```
Comment (abaranch) [inline test/SampleWebAPIIntegration/Startup.cs]: You want to uncomment this

File: test/SampleWebAPIIntegration/Startup.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            // Add Application Insights exceptions handling to the request pipeline.
+            app.UseApplicationInsightsExceptionTelemetry();
+
+            // Configure the HTTP request pipeline.
+            app.UseStaticFiles();
+
+            // Add MVC to the request pipeline.
+            app.UseMvc();
+            // Add the following route for porting Web API 2 controllers.
+            // routes.MapWebApiRoute("DefaultApi", "api/{controller}/{id?}");
```
Comment (abaranch) [inline test/SampleWebAPIIntegration/Startup.cs]: You want to uncomment this

File: test/WebApplication1/Controllers/ValuesController.cs (hunk 1 above)
Comment (abaranch) [inline test/WebApplication1/Controllers/ValuesController.cs]: Can we rename current sample app that we have to Mvc6Framework45TestApp and this one to WebApiShimFw46TestApp or similar to reflect what is what? (I suppose you used a different template, different fw version and you will uncomment web api shim so we can test request names)

Comment (SergKanz) [inline test/WebApplication1/Controllers/ValuesController.cs]: Renamed

File: test/SampleWebAPIIntegration/Project_Readme.html (hunk 2 above)
Comment (SergKanz) [inline test/SampleWebAPIIntegration/Project_Readme.html]: Deleted both Readme files

File: test/SampleWebAPIIntegration/Controllers/ValuesController.cs (hunk 3 above)
Comment (SergKanz) [inline test/SampleWebAPIIntegration/Controllers/ValuesController.cs]: We need to cover these cases in Unit tests. Functional tests are just integration test

File: test/SampleWebAPIIntegration/SampleWebAPIIntegration.xproj (hunk 4 above)
Comment (SergKanz) [inline test/SampleWebAPIIntegration/SampleWebAPIIntegration.xproj]: Done

File: test/SampleWebAPIIntegration/FunctionalTest/RequestTelemetryTests.cs (hunk 5 above)
Comment (SergKanz) [inline test/SampleWebAPIIntegration/FunctionalTest/RequestTelemetryTests.cs]: sure

File: test/SampleWebAPIIntegration/Startup.cs (hunk 6 above)
Comment (SergKanz) [inline test/SampleWebAPIIntegration/Startup.cs]: aha, makes sense to test both routes

File: test/SampleWebAPIIntegration/Startup.cs (hunk 7 above)
Comment (SergKanz) [inline test/SampleWebAPIIntegration/Startup.cs]: Done

**Comment by SergKanz** [comment]
ping: @abaranch @olegsych

**Comment by abaranch** [comment]
Approved

**Comment by olegsych** [comment]
:+1:

File: test/SampleWebAPIIntegration/FunctionalTest/RequestTelemetryTests.cs (hunk 5 above)
Comment (olegsych) [inline test/SampleWebAPIIntegration/FunctionalTest/RequestTelemetryTests.cs]: So, we now have two functional test classes for request tracking - one for [web api](https://github.com/Microsoft/ApplicationInsights-aspnetv5/blob/develop/test/WebApiShimFw46.FunctionalTests/FunctionalTest/RequestTelemetryTests.cs) and one for [Mvc](https://github.com/Microsoft/ApplicationInsights-aspnetv5/blob/develop/test/Mvc6Framework45.FunctionalTests/FunctionalTest/RequestTelemetryTests.cs) and *no* unit tests for the request tracking middleware. How would you evaluate this implementation against the [four rules of simple design](http://martinfowler.com/bliki/BeckDesignRules.html)?

cc: @SergKanz, @abaranch, @upendras, @Sergei-Nikitin

**Pull request merged**